### PR TITLE
[WIP]EOS 19219: btree benchmark framework

### PR DESCRIPTION
**Pull request by NikhilBirgade**

### CODE REVIEW

File: motr/m0crate/tests/test_btree.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#
+# Copyright (c) 2020 Seagate Technology LLC and/or its Affiliates
```
Comment (shashank-parulekar) [inline motr/m0crate/tests/test_btree.yaml:2]: 2021 !!!

**Review by shashank-parulekar — CHANGES_REQUESTED** [review]

File: motr/m0crate/crate.c [hunk 2]
```diff
@@ -1152,6 +1186,236 @@ static void csum_check (struct workload *w)
...
+	return 0;
+}
+
+static void btree_run(struct workload *w, struct workload_task *task)
+{
+	int                       i;
+	int                       sum = 0;
+	struct cr_workload_btree *cwb = w->u.cw_btree;
+
+	cr_log(CLL_INFO, "Total ops:             %i\n", w->cw_ops);
```
[the remark this answers is not in the document]
Comment (NikhilBirgade) [reply motr/m0crate/crate.c:1257]: Done.

File: motr/m0crate/crate.c [hunk 3]
```diff
@@ -1152,6 +1186,236 @@ static void csum_check (struct workload *w)
...
+	cr_log(CLL_INFO, "key size:              %i\n", cwb->cwb_key_size);
+	cr_log(CLL_INFO, "value size:            %i\n", cwb->cwb_value_size);
+	cr_log(CLL_INFO, "max key size:          %i\n", cwb->cwb_max_key_size);
+	cr_log(CLL_INFO, "max value size:        %i\n", cwb->cwb_max_value_size);
+	cr_log(CLL_INFO, "keys order:            %i\n", cwb->cwb_keys_ordered);
```
[the remark this answers is not in the document]
Comment (NikhilBirgade) [reply motr/m0crate/crate.c:1262]: ok

File: motr/m0crate/crate.c [hunk 4]
```diff
@@ -1152,6 +1186,236 @@ static void csum_check (struct workload *w)
...
+	}
+
+	M0_ALLOC_PTR(kv);
+	m0_buf_init(&kv->kv_key, k, ksize);
+	m0_buf_init(&kv->kv_val, v, vsize);
+
+	//TODO: Add code to append pattern cwb->cwb_pattern
+	if (cwb->cwb_keys_ordered) {
+		cwb->cwb_bo[ot].key = cwb->cwb_bo[ot].key + 1 >
+				      cwb->cwb_bo[ot].nr_ops ?
```
[the remark this answers is not in the document]
Comment (NikhilBirgade) [reply motr/m0crate/crate.c:1373]: Thanks. Earlier I was calculating cwb->cwb_bo[ot].nr_ops during init(based on percentage) and taking that many ops per operation to wrap up. Later I changed logic to increase  cwb->cwb_bo[ot].nr at the end of ops execution. I will change the logic to wrap up keys.

File: motr/m0crate/tests/test_btree.yaml (hunk 1 above)
Comment (NikhilBirgade) [reply motr/m0crate/tests/test_btree.yaml:2]: Thanks

**Review by ydb242 — APPROVED** [review]
Looks good to me!

File: motr/m0crate/tests/test_btree.yaml (hunk 1 above)
Comment (shashank-parulekar) [reply motr/m0crate/tests/test_btree.yaml]: You have only changed the copyright notice in this yaml file please change in the rest of the modified 7 files in this PR.

File: motr/m0crate/crate.c [hunk 5]
```diff
@@ -1152,6 +1186,269 @@ static void csum_check (struct workload *w)
...
+		case BOT_DELETE:
+			cr_btree_delete(&kv);
+			break;
+		default:
+			break;
+	        }
+
+	etime = m0_time_sub(m0_time_now(), stime);
+	pthread_mutex_lock(&w->cw_lock);
+	cr_log(CLL_TRACE, "op:%s key=%.*s%"PRIu64" ksize=%d val=%s vsize=%d\n",
```
[the remark this answers is not in the document]
Comment (NikhilBirgade) [reply motr/m0crate/crate.c:1433]: Ok.

File: motr/m0crate/crate.c [hunk 6]
```diff
@@ -1152,6 +1186,269 @@ static void csum_check (struct workload *w)
...
+	memset(kpattern, pattern, psize);
+	m0_bitstring_copy(&bk->pattern, kpattern, psize);
+	bk->bkey = key;
+}
+
+static void btree_op_run(struct workload *w, struct workload_task *task,
+			 const struct workload_op *op)
+{
+	m0_time_t                 stime;
+	m0_time_t                 etime;
```
[the remark this answers is not in the document]
Comment (NikhilBirgade) [reply motr/m0crate/crate.c:1383]: Ok.

**Pull request merged**